### PR TITLE
Added row selection feature

### PR DESCRIPTION
**Pull request by 3rfan**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| License | MIT

Related PR https://github.com/sulu/sulu/pull/3497

#### What's in this PR?

Row selection for the table component. 

#### Why?

Row selection is needed.

#### To Do

- [x] Create documentation
- [x] Write tests
- [x] Add Checkbox component
- [x] Add Radio component

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 1]
```diff
@@ -1,21 +1,151 @@
...
 import type {ChildrenArray, Element} from 'react';
 import React from 'react';
+import Icon from '../Icon';
+import Cell from './Cell';
 import Row from './Row';
+import type {ControlItems, ControlConfig, RowProps, SelectMode} from './types';
+import tableStyles from './table.scss';
 
 type Props = {
+    /** Child nodes of the table body */
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Remove

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 2]
```diff
@@ -1,21 +1,151 @@
...
+
+    isSelectable = () => {
+        const {selectMode} = this.props;
+
+        return selectMode === 'single' || selectMode === 'multiple';
+    };
+
+    cloneRows = (originalRows: ChildrenArray<Element<typeof Row>>) => {
+        return React.Children.map(originalRows, (row, index) => {
+            if (React.isValidElement(row)) {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: We usually try to do early returns, since it keeps the level of indentation down. And why is this line necessary at all?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 3]
```diff
@@ -1,21 +1,151 @@
...
+                    },
+                    cells,
+                );
+            }
+        });
+    };
+
+    createCells = (cells: ChildrenArray<Element<typeof Cell>>, rowProps: RowProps, rowIndex: number) => {
+        const {controls} = this.props;
+        const prependedCells = [];
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Not sure if I would solve it like this or via props in the `Row`.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js [hunk 4]
```diff
@@ -4,6 +4,10 @@ import tableStyles from './table.scss';
 import type {RowProps} from './types';
 
 export default class Row extends React.PureComponent<RowProps> {
+    static defaultProps = {
+        selected: false,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: Shouldn't the row also render the checkbox itself, if you are passing the properties this way?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js [hunk 5]
```diff
@@ -1,21 +1,105 @@
...
+    selectMode?: SelectMode,
+    /** 
+     * Callback function to notify about selection and deselection of a row.
+     * If the "id" prop is set on the row, the "rowId" corresponds to that, else it is the index of the row.
+     */
+    onRowSelectionChange?: (rowId: string | number, selected: boolean) => void,
+    /** Called when the "select all" checkbox in the header was clicked. Returns the checked state. */
+    onSelectAllChange?: (checked: boolean) => void,
+    /** If true the "select all" checkbox is checked. */
+    selectAllChecked?: boolean,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js]: This shouldn't be necessary, but we've already talked about that :-)

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js (hunk 2 above)
Comment (3rfan) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Verifies the object is a React element. Returns true or false.

https://facebook.github.io/react/docs/react-api.html#isvalidelement

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/HeaderCell.js [hunk 6]
```diff
@@ -1,17 +1,77 @@
...
     children?: Node,
+    /**
+     * @ignore 
+     * If set to true, the cell appears as a header-control-cell
+     */
+    isControl?: boolean,
+    /** Called when column was clicked */
+    onClick?: () => void,
+    /** If set, an indicator will show up */
+    sortMode: 'default' | 'ascending' | 'descending',
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/HeaderCell.js]: Wouldn't `none` be more appropriate than `default`? It's not sorted at all by this column in most cases.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/README.md [hunk 7]
```diff
@@ -73,3 +73,297 @@ const HeaderCell = require('./HeaderCell').default;
     </Body>
 </Table>
 ```
+
+An empty table components will be rendered as follows. You can set a `placeholderText`.
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/README.md]: `component`

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js (hunk 4 above)
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: Is it possible that this prop is not needed anymore?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js [hunk 8]
```diff
@@ -1,22 +1,136 @@
...
 
+@observer
 export default class Table extends React.PureComponent<Props> {
+    static defaultProps = {
+        selectMode: 'none',
+    };
+
+    @observable allSelected = false;
+
+    @action setAllSelected = (allSelected: boolean) => {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js]: I think it is weird that this is a action... Why not just checking all rows and infer it from there? This way it might just end up having the wrong value...

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss [hunk 9]
```diff
@@ -78,4 +108,50 @@ $cellBackgroundColorActive: $shakespeare;
         border-radius: 0 3px 3px 0;
         border-right-width: 1px;
     }
+
+    &.small {
+        width: 1%;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss]: why `1%` wouldn't `1px` be the better choice? (Would be smaller or equal than the actual width for sure)

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss [hunk 10]
```diff
@@ -78,4 +108,50 @@ $cellBackgroundColorActive: $shakespeare;
...
+    }
+}
+
+.control-cell {
+    padding: 0;
+    background-color: $cellBackgroundColor;
+    color: $cellTextColorActive;
+    text-align: center;
+    width: 1%;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss]: `1px`?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss [hunk 11]
```diff
@@ -78,4 +108,50 @@ $cellBackgroundColorActive: $shakespeare;
...
+        width: 55px;
+        height: 38px;
+        border: none;
+        color: inherit;
+        cursor: pointer;
+        background-color: transparent;
+    }
+}
+
+.table-placeholder-area {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss]: Why don't we have a `table-placeholder` class with nested `area`, `icon` and `text` classes, instead of three prefixed classes?

Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/table.scss]: nesting increases the specificity of a selector. We should only use that for state specific selectors (e.g. small, hidden, etc.).

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js [hunk 12]
```diff
@@ -1,22 +1,151 @@
...
+    };
+
+    isSingleSelect = () => {
+        return this.props.selectMode === 'single';
+    };
+
+    createHeaderRow = (originalRows: any) => {
+        const rows = React.Children.toArray(originalRows);
+
+        if (rows.length > 1) {
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js]: Tried that but got an error message :/

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 13]
```diff
@@ -25,163 +21,39 @@ type Props = {
...
+            return React.cloneElement(
+                row,
+                {
+                    ...row.props,
+                    key: `body-row-${index}`,
+                    rowIndex: index,
+                    buttons: buttons,
+                    selectMode: selectMode,
+                    onSingleSelectionChange: this.handleRowSingleSelectionChange,
+                    onMultipleSelectionChange: this.handleRowMultipleSelectionChange,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Can't the previous two be combined into one somehow?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js [hunk 14]
```diff
@@ -1,21 +1,24 @@
...
+    /**
+     * @ignore
      * List of buttons to apply action handlers to every row (e.g. edit row).
-     * The header will display the icons.
+     * The header will just display the icons.
+     */
+    buttons?: Array<ButtonConfig>,
+    /** 
+     * @ignore
+     * Can be set to "single" or "multiple". Defaults is "none".
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js]: `Default`

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js (hunk 13 above)
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: I could probably do that by also setting the first function argument of the single select to a boolean (checked). But idk if that would be a good solution

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 15]
```diff
@@ -1,11 +1,10 @@
 // @flow
-import type {ChildrenArray, Element} from 'react';
+import type {ChildrenArray} from 'react';
 import React from 'react';
-import Row from './Row';
 import type {ButtonConfig, SelectMode} from './types';
 
 type Props = {
-    children?: ChildrenArray<Element<typeof Row>>,
+    children?: ChildrenArray<*>,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Doesn't the `Body` only allow `Row`s as children? Then this would be wrong...

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js [hunk 16]
```diff
@@ -8,7 +8,7 @@ import type {ButtonConfig, SelectMode} from './types';
 import tableStyles from './table.scss';
 
 type Props = {
-    children: ChildrenArray<Element<typeof HeaderCell>>,
+    children: ChildrenArray<*>,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Header.js]: Doesn't the `Header` only allow `HeaderCell`s as children?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js [hunk 17]
```diff
@@ -1,13 +1,39 @@
...
 import {Radio} from '../Radio';
-import type {ButtonConfig, RowProps} from './types';
+import type {ButtonConfig, SelectMode} from './types';
 import Cell from './Cell';
 import ButtonCell from './ButtonCell';
 import tableStyles from './table.scss';
 
-export default class Row extends React.PureComponent<RowProps> {
+type Props = {
+    children: ChildrenArray<*>,
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Row.js]: Shouldn't we be able to type this more specific? It only allows certain types of `Cell`s, doesn't it?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js [hunk 18]
```diff
@@ -33,32 +33,11 @@ export default class Table extends React.PureComponent<Props> {
...
-                    throw new Error(
-                        'The Table component only accepts the following children types: ' +
-                        [Header.name, Body.name].join(', ')
-                    );
-            }
-        });
-
-        return {body, header};
-    };
+    cloneHeader = (originalHeader?: any, allSelected: boolean) => {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js]: Can't we have the correct header type here instead of `any`?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js [hunk 19]
```diff
@@ -70,7 +49,11 @@ export default class Table extends React.PureComponent<Props> {
         );
     };
 
-    cloneBody = (originalBody: any) => {
+    cloneBody = (originalBody?: any) => {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Table.js]: Can't we have the correct body type here instance of any?

**Comment by danrot** [comment]
@3rfan Here you go, just 3 `any`s, which can be removed with the next version of flow 😉

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js (hunk 15 above)
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: Everyone of the mentioned components except the table is throwing flow errors when the children type is set. I guess it's because those components are inside React.Children.map/forEach. `ChildrenArray<*>` is a bit more specific than `any`.

Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: You know that I've fixed that without `*` or `any` in the latest commit in this PR?

**Comment by 3rfan** [comment]
Ok thx, the errors probably came from not using `?` before the `ChildrenArray` type. Is the PR now ready to review by @chirimoya ?

**Comment by danrot** [comment]
Have overseen one more occurence, now it should be find and can be reviewed.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js [hunk 20]
```diff
@@ -1,21 +1,61 @@
...
 import React from 'react';
+import type {ButtonConfig, SelectMode} from './types';
 import Row from './Row';
 
 type Props = {
-    children: ChildrenArray<Element<typeof Row>>,
+    children?: ChildrenArray<Element<typeof Row>>,
+    /**
+     * @ignore
+     * List of buttons to apply action handlers to every row (e.g. edit row) forwarded from table
```
Comment (chirimoya) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Table/Body.js]: remote duplicate comments
